lib/system/environ/validate.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package environ implements utilities for managing host environment
+package environ
+
+import (
+	"os"
+	"strings"
+
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/localenv"
+	"github.com/gravitational/gravity/lib/pack"
+	"github.com/gravitational/gravity/lib/state"
+	"github.com/gravitational/gravity/lib/system"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+)
+
+// ValidateInstall performs a local environment sanity check to make sure
+// that install on this node can proceed without issues
+func ValidateInstall(env *localenv.LocalEnvironment) func() error {
+	return func() error {
+		if err := validateNonVolatileDirectory(state.GravityInstallDir()); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := validateDirectoryEmpty(state.GravityInstallDir()); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := validateNoPackageState(env.Packages, env.StateDir); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	}
+}
+
+func validateNonVolatileDirectory(stateDir string) error {
+	fstype, err := system.GetFilesystemForPath(stateDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if fstype == system.FilesystemTemporary {
+		return trace.BadParameter("installer is running from a temporary file system.\n" +
+			"It is required to run the installer from a non-volatile location.")
+	}
+	var volatileDirectories = []string{"/tmp", "/var/tmp"}
```

Comment:
I'm pretty sure this will break installs for half of our customers :) I.e. I know for sure two our major customers run installers from somewhere in /tmp. Where did this requirement come from?

Reply:
This requirement comes from the fact that in order to support resumption after the node is rebooted, the state needs to reside somewhere non-volatile. I was already thinking of making this a warning instead of the hard requirement as installs can and do complete without the node being rebooted.

Reply:
Regarding the choice of the operation state directory - taking from git's approach to storing state in a local .git directory which I think is very flexible.
The fact that it forces the user to execute the binary from the same directory is not a disadvantage - contrast this with the requirement that update places on its operation. Why should this be different for an install or join operation?
Additionally, using a different location per operation is not ideal and leads to proliferation of state which we have to deal with when we try to compute the momentary operational state of the cluster - why do we need to do this if there's just one operation in flight at any given time? Why not keep cluster state in `--state-dir` and any active operation state in `some-other-state-dir` (which would be the same for any operation) so we don't need to consult 3-4 places to build the state view.

Having the temporary state directory written to a central directory (i.e. `/etc/.gravity`) as opposed to local directory relative to where the application state is (i.e. unpacked data for install) is fragile to updates if we need to differentiate the location between install/join/update/update_config/update_environ/etc operation state directories.